File: java/src/main/java/shapes/base/models/Square.java
package shapes.base.models;

import shapes.base.commons.Dimensional;

public final class Square implements Dimensional {
    private double side;

    public Square(double side) {
        double dimOne = (side > 0) ? side : 1;
        this.side = dimOne;
    }

    @Override
    public double area() {
        return Math.pow(this.side, 2);
    }

    @Override
    public double perimeter() {
        return 4 * side;
    }
}

File: elixir/lib/shapes/base/models/square.ex
alias Shapes.Base.Commons.Dimensional
alias Shapes.Base.Models.Square

defmodule Square do
  defstruct side: 0

  def new(side) do
    dim_one = if(side > 0, do: side, else: 1.0)
    %__MODULE__{side: dim_one}
  end

  def area(self) do
    Dimensional.area(self)
  end

  def perimeter(self) do
    Dimensional.perimeter(self)
  end
end

defimpl Dimensional, for: Square do
  def area(%Square{side: side}) do
    :math.pow(side, 2)
  end

  def perimeter(%Square{side: side}) do
    4 * side
  end
end
